### PR TITLE
Support multiple files in AST

### PR DESCRIPTION
**Pull request by bthesorceror**

### CODE REVIEW

File: src/shared/mod.rs [hunk 1]
```diff
@@ -94,11 +93,57 @@ pub fn apply_changes(
...
+            utils::create_file_node_from_text(
+                v.uri.clone(),
+                v.contents,
+            )
+        })
+        .collect::<Vec<flux::ast::Package>>();
+
+    let pkg = pkgs.into_iter().fold(
+        None,
+        |acc: Option<flux::ast::Package>, pkg| {
```
Comment (jsteenb2) [inline src/shared/mod.rs]: I'm curious, why is the `Option` needed here?

Comment (bthesorceror) [reply src/shared/mod.rs]: because we need to start at a state of None as its possible that get_package will return no values, we build off the first pkg after it is confimed there is a first one.

File: src/visitors/semantic/completion/mod.rs [hunk 2]
```diff
@@ -192,15 +192,19 @@ impl<'a> Visitor<'a> for CompletableObjectFinderVisitor {
 }
 
 fn get_var_type(expr: &Expression) -> Option<VarType> {
+    match expr.type_of() {
```
Comment (jsteenb2) [inline src/visitors/semantic/completion/mod.rs]: also curious, why was this needed here?

File: src/shared/mod.rs (hunk 1 above)
Comment (jsteenb2) [reply src/shared/mod.rs]: ahh ok, thanks 👍

File: src/visitors/semantic/completion/mod.rs (hunk 2 above)
Comment (bthesorceror) [reply src/visitors/semantic/completion/mod.rs]: in the case of something like `-1h` this was actually showing up as a unary expression rather than a duration but if we use the type inference from flux we can determine the type.

Comment (jsteenb2) [reply src/visitors/semantic/completion/mod.rs]: ahh nice find, the remaining types in the match below this, are we not able to use the `type_of()` for them? curious why there is a split 🤔

Comment (bthesorceror) [reply src/visitors/semantic/completion/mod.rs]: they probably can but I want to do that in a different PR

Comment (bthesorceror) [reply src/visitors/semantic/completion/mod.rs]: the remaining ones will be a larger change

File: src/visitors/semantic/utils.rs [hunk 3]
```diff
@@ -81,14 +81,16 @@ pub fn create_completion_package_removed(
         .filter(|x| valid_node(x, pos.clone()))
         .collect();
 
-    let ast_pkg = flux::ast::Package {
-        base: file.base.clone(),
-        path: "".to_string(),
-        package: "main".to_string(),
-        files: vec![file],
-    };
+    let mut pkg = crate::shared::create_ast_package(uri.clone())?;
```
Comment (jsteenb2) [inline src/visitors/semantic/utils.rs]: can you add a test to verify completion works with multiple files?

File: src/visitors/semantic/completion/mod.rs (hunk 2 above)
Comment (jsteenb2) [reply src/visitors/semantic/completion/mod.rs]: right on, thanks for sharing 👍

File: src/cache/mod.rs [hunk 4]
```diff
@@ -5,6 +5,27 @@ lazy_static! {
     static ref GLOBAL_CACHE: Cache = Cache::default();
 }
 
+fn get_dir(uri: String) -> String {
+    let mut parts = uri.split('/').collect::<Vec<&str>>();
+    parts.pop();
+    parts.join("/")
```
Comment (kelwang) [inline src/cache/mod.rs]: This could build a lot memory allocation. Should use [rfind](https://doc.rust-lang.org/stable/std/primitive.str.html#method.rfind) , to get the last appearance of the string, and use substring to get the result.

File: src/cache/mod.rs [hunk 5]
```diff
@@ -5,6 +5,27 @@ lazy_static! {
     static ref GLOBAL_CACHE: Cache = Cache::default();
 }
 
+fn get_dir(uri: String) -> String {
```
Comment (kelwang) [inline src/cache/mod.rs]: can you use &str for uri and return?

File: src/shared/mod.rs [hunk 6]
```diff
@@ -94,11 +100,60 @@ pub fn apply_changes(
     original
 }
 
+pub fn create_ast_package(
+    uri: String,
+    ctx: RequestContext,
+) -> Result<flux::ast::Package, String> {
+    let values =
+        cache::get_package(uri.clone(), ctx.support_multiple_files)?;
```
Comment (kelwang) [inline src/shared/mod.rs]: The get_package will require all the files in the package already opened, then they can be cached right? What if a file is not opened, will the parser returns some kind of error?

Comment (bthesorceror) [reply src/shared/mod.rs]: This is why we are going to disable it in vsflux for the time being, we will need to figure out the best way to prime the cache.

File: src/cache/mod.rs (hunk 4 above)
Comment (jademcgough) [reply src/cache/mod.rs]: I think this implementation is fine - the memory allocation is small, this is a pretty common pattern for parsing directories, and it favors readability.
